module/functions/azure/aad/Get-AzureAdDirectoryObject.Tests.ps1

```diff
@@ -0,0 +1,202 @@
+    Mock _groupByName {}
+    Mock _spByAppId { $global:methodUsed="ApplicationId"; $mockServicePrincipal } -ParameterFilter { $ApplicationId -eq $mockServicePrincipal.ApplicationId }
+    Mock _spByAppId {}
+    Mock _spByObjectId { $global:methodUsed="ObjectId"; $mockServicePrincipal } -ParameterFilter { $ObjectId -eq $mockServicePrincipal.ObjectId }
+    Mock _spByObjectId {}
+    Mock _spByName { $global:methodUsed="DisplayName"; $mockServicePrincipal } -ParameterFilter { $DisplayName -eq $mockServicePrincipal.DisplayName }
+    Mock _spByName {}
+    Mock _userById { $global:methodUsed="ObjectId"; $mockUser } -ParameterFilter { $ObjectId -eq $mockUser.ObjectId }
+    Mock _userById {}
+    Mock _userByName { $global:methodUsed="DisplayName"; $mockUser } -ParameterFilter { $DisplayName -eq $mockUser.DisplayName }
+    Mock _userByName {}
+    Mock _userByUpn { $global:methodUsed="UserPrincipalName"; $mockUser } -ParameterFilter { $UserPrincipalName -eq $mockUser.UserPrincipalName }
+    Mock _userByUpn {}
+
+    Mock Write-Verbose {}
+    Mock Write-Warning {}
+
+    Context "Finding a group" {
+
+        Context "Searching by ObjectId" {
+            $global:methodUsed = $null
+            $res = Get-AzureAdDirectoryObject -Criterion $mockGroup.ObjectId
+            
+            It "should return the group" {
+                $methodUsed | Should -Be "ObjectId"
+                $res.DisplayName | Should -Be $mockGroup.DisplayName
+                
+                Assert-MockCalled _groupById -Times 1
+                Assert-MockCalled _spByAppId -Times 1
+                Assert-MockCalled _spByObjectId -Times 1
+                Assert-MockCalled _userById -Times 1
+                Assert-MockCalled _groupByName -Times 0
+                Assert-MockCalled _spByName -Times 0
+                Assert-MockCalled _userByName -Times 0
+            }
+        }
+
+        Context "Searching by DisplayName" {
+            $global:methodUsed = $null
+            $res = Get-AzureAdDirectoryObject -Criterion $mockGroup.DisplayName
+            
+            It "should return the group" {
+                $methodUsed | Should -Be "DisplayName"
+                $res.ObjectId | Should -Be $mockGroup.ObjectId
+                
+                Assert-MockCalled _groupById -Times 0
+                Assert-MockCalled _spByAppId -Times 0
+                Assert-MockCalled _spByObjectId -Times 0
+                Assert-MockCalled _userById -Times 0
+                Assert-MockCalled _groupByName -Times 1
+                Assert-MockCalled _spByName -Times 1
+                Assert-MockCalled _userByName -Times 1
+            }
+        }
+    }
+
+    Context "Finding a service principal" {
+
+        Context "Searching by ObjectId" {
+            $global:methodUsed = $null
+            $res = Get-AzureAdDirectoryObject -Criterion $mockServicePrincipal.ObjectId
+            
+            It "should return the group" {
```

Comment:
Why should "finding a service principal" return a group?

Reply:
copy/paste FTW! Updated.